packages/gensx/tests/checkpoint.test.tsx

```diff
@@ -0,0 +1,359 @@
+
+    // Verify checkpoint structure
+    expect(finalCheckpoint).toMatchObject({
+      componentName: "SimpleComponent",
+      props: { message: "world" },
+      output: "hello world",
+    });
+
+    // Verify timing fields
+    expect(finalCheckpoint.startTime).toBeDefined();
+    expect(finalCheckpoint.endTime).toBeDefined();
+    expect(finalCheckpoint.startTime).toBeLessThan(finalCheckpoint.endTime!);
+  });
+
+  test("no checkpoints when disabled", async () => {
+    // Disable checkpoints
+    process.env.GENSX_CHECKPOINTS = undefined;
+
+    // Define a simple component that returns a string
+    const SimpleComponent = gsx.Component<{ message: string }, string>(
+      "SimpleComponent",
+      async ({ message }) => {
+        await setTimeout(0);
+        return `hello ${message}`;
+      },
+    );
+
+    // Execute workflow and get results
+    const { result, checkpoints } = await executeWithCheckpoints<string>(
+      <SimpleComponent message="world" />,
+    );
+
+    // Verify execution still works
+    expect(result).toBe("hello world");
+
+    // Verify no checkpoint calls were made
+    expect(global.fetch).not.toHaveBeenCalled();
+    expect(checkpoints).toHaveLength(0);
+  });
+
+  test("handles parallel execution", async () => {
+    // Define a simple component that we'll use many times
+    const SimpleComponent = gsx.Component<{ id: number }, string>(
+      "SimpleComponent",
+      async ({ id }) => {
+        await setTimeout(0); // Small delay to ensure parallel execution
+        return `component ${id}`;
+      },
+    );
+
+    // Create a component that returns an array of parallel executions
+    const ParallelComponent = gsx.Component<{}, string[]>(
+      "ParallelComponent",
+      async () => {
+        return Promise.all(
+          Array.from({ length: 100 }).map((_, i) =>
+            gsx.execute<string>(<SimpleComponent id={i} />),
+          ),
+        );
+      },
+    );
+
+    // Execute workflow and get results
+    const { result, checkpoints } = await executeWithCheckpoints<string[]>(
+      <ParallelComponent />,
+    );
+
+    // Verify execution result
+    expect(result).toHaveLength(100);
+    expect(result[0]).toBe("component 0");
+    expect(result[99]).toBe("component 99");
+
+    // Verify checkpoint behavior
+    const fetchCalls = (global.fetch as unknown as ReturnType<typeof vi.fn>)
+      .mock.calls.length;
+
+    // We expect:
+    // - Some minimum number of calls to capture the state (could be heavily batched)
+    // - Less than the theoretical maximum (303 = parent(2) + children(200) + execute calls(101))
+    // - Evidence of queueing (significantly less than theoretical maximum)
+    expect(fetchCalls).toBeGreaterThan(0); // At least some calls must happen
+    expect(fetchCalls).toBeLessThan(303); // Less than theoretical maximum
+    expect(fetchCalls).toBeLessThan(250); // Evidence of significant queueing
```

Comment:
Any concerns about this being flakey?

Reply:
303 is the max number of calls it can theoretically make. 

everytime I've run it it has been exactly 204 (or maybe 205, I can't remember) - but if it turns out to be flakey we can remove the middle constraint.